File: frontend/lib/conversion.ts
export type ConversionStatus = "completed" | "failed";

export interface OrientedBoundingBox {
  centerX: number;
  centerY: number;
  width: number;
  height: number;
  rotation: number;
}

export interface DetectedSymbol {
  id: string;
  label: string;
  confidence: number;
  boundingBox: OrientedBoundingBox;
  pageNumber: number;
}

export interface PdfPageResult {
  pageNumber: number;
  width: number;
  height: number;
  detectedSymbols: DetectedSymbol[];
}

export interface PatternInstruction {
  id: string;
  row: number;
  text: string;
  confidence: number | null;
}

export interface ConversionResponse {
  status: ConversionStatus;
  mock: boolean;
  title: string;
  instructions: PatternInstruction[];
  pages: PdfPageResult[];
  averageConfidence: number | null;
  message?: string;
}

interface ApiErrorBody {
  detail?: string;
}

const API_URL = (process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000").replace(/\/$/, "");

export async function convertPattern(file: File): Promise<ConversionResponse> {
  const formData = new FormData();
  formData.append("file", file);

  let response: Response;
  try {
    response = await fetch(`${API_URL}/api/convert`, {
      method: "POST",
      body: formData,
    });
  } catch {
    throw new Error("Could not reach the conversion service. Make sure the FastAPI server is running and try again.");
  }

  if (!response.ok) {
    const body = await readErrorBody(response);
    throw new Error(body?.detail || `Conversion failed with status ${response.status}.`);
  }

  const result: unknown = await response.json();
  if (!isConversionResponse(result)) {
    throw new Error("The conversion service returned an unexpected response.");
  }

  return result;
}

async function readErrorBody(response: Response): Promise<ApiErrorBody | null> {
  try {
    return (await response.json()) as ApiErrorBody;
  } catch {
    return null;
  }
}

function isConversionResponse(value: unknown): value is ConversionResponse {
  if (!value || typeof value !== "object") return false;
  const response = value as Partial<ConversionResponse>;
  return (
    (response.status === "completed" || response.status === "failed") &&
    typeof response.mock === "boolean" &&
    typeof response.title === "string" &&
    Array.isArray(response.instructions) &&
    Array.isArray(response.pages) &&
    (typeof response.averageConfidence === "number" || response.averageConfidence === null)
  );
}

File: backend/app/schemas/conversion.py
from typing import Literal

from pydantic import BaseModel, Field


class HealthResponse(BaseModel):
    status: Literal["ok"] = "ok"
    service: str = "crochet-pattern-converter"


class OrientedBoundingBox(BaseModel):
    centerX: float = Field(ge=0, le=1)
    centerY: float = Field(ge=0, le=1)
    width: float = Field(gt=0, le=1)
    height: float = Field(gt=0, le=1)
    rotation: float


class DetectedSymbol(BaseModel):
    id: str
    label: str
    confidence: float = Field(ge=0, le=1)
    boundingBox: OrientedBoundingBox
    pageNumber: int = Field(ge=1)


class PdfPageResult(BaseModel):
    pageNumber: int = Field(ge=1)
    width: int = Field(gt=0)
    height: int = Field(gt=0)
    detectedSymbols: list[DetectedSymbol]


class PatternInstruction(BaseModel):
    id: str
    row: int = Field(ge=1)
    text: str
    confidence: float | None = Field(default=None, ge=0, le=1)


class ConversionResponse(BaseModel):
    status: Literal["completed", "failed"]
    mock: bool
    title: str
    instructions: list[PatternInstruction]
    pages: list[PdfPageResult]
    averageConfidence: float | None = Field(default=None, ge=0, le=1)
    message: str | None = None
